### PR TITLE
feat(institution-admin): add access revocation and TOTP reset (#355)

### LINKED ISSUE #355: feat(institution-web): Institution Phase F — Account Recovery + Hard Revocation
## Summary

Phase F addresses two institution account integrity gaps that are real pilot/operational risks: the absence of a robust account recovery path for institution users who lose TOTP device access, and the absence of a hard revocation mechanism for removing institution access. Today, institution access can only be demoted in role, not truly revoked. Phase F defines, implements, and enforces both flows end-to-end across backend, frontend, and audit trail.

---

## Why this phase exists now

The post-Phase A–D audit surfaced two operationally critical gaps:

1. **No recovery path**: an institution user who loses their TOTP device has no supported recovery route. This is a real pilot risk — a single device loss strands an operator with no recourse.
2. **No hard revocation**: institution admins can demote a user to a lower role, but there is no clean "remove access entirely" action. A demoted user still has a valid account. This leaves a blurry boundary between "reduced access" and "no access."

Both gaps must be addressed before pilot, where real institutional staff will use the product under real operational conditions.

---

## Current state at HEAD

**Exists:**
- TOTP setup and enrollment (PR #335)
- Role management: invite (operator/admin), promote/demote via `PUT /v1/institution-admin/users/{userId}/role`
- Step-up session gating on admin write mutations
- Audit logging for invite and role-change actions (PR #350)
- Institution membership model with `institution` / `institution_admin` roles

**Missing:**
- TOTP recovery path (backup codes, admin-initiated reset, or documented recovery procedure)
- Hard revocation endpoint and model: a distinct "remove access" state separate from role demotion
- Backend enforcement of revoked-user behavior (what can a revoked account do?)
- Admin UX for triggering recovery-related actions
- Admin UX for hard revocation with confirmation
- Tests covering recovery and revocation edge cases

---

### Backend
- Design and implement the account recovery model:
  - Option A: TOTP reset by institution admin — admin action clears the user's TOTP registration, user must re-enroll on next login
  - Option B: backup codes generated at TOTP enrollment — user can use a one-time backup code to bypass TOTP
  - Evaluate and choose one; document the decision in the PR body
- Implement hard revocation:
  - Distinct `revoked` state on institution membership (not just role = viewer)
  - Revoked user cannot authenticate into institution surfaces
  - Revoked user receives a clear error on attempted login
  - `DELETE /v1/institution-admin/users/{userId}` or equivalent endpoint for hard revocation
- Audit trail: both recovery and revocation actions must emit structured log entries and audit records
- Schema migration if new columns or states are needed

### Frontend
- Admin UX: "Reset TOTP" or "Initiate Recovery" action (visible to institution_admin only)
- Admin UX: "Revoke Access" action with confirmation modal (visible to institution_admin only)
- Confirmation flow for revocation (destructive action — require explicit confirmation)
- Correct error UX for a revoked user attempting to use the dashboard
- Step-up session gate on both recovery-trigger and revocation actions

### Contracts / docs
- New or modified endpoints must be reflected in `docs/arch/02_api_contracts.md`
- OpenAPI spec (`02_openapi.yaml`) updated with any new endpoints
- `AUTHORIZATION_COVERAGE_MATRIX.md` updated if new protected routes are added

### Tests
- Test: admin can trigger TOTP reset for a user (or equivalent recovery action)
- Test: user after TOTP reset must re-enroll before accessing institution surfaces
- Test: admin can hard-revoke a user's access
- Test: revoked user receives correct rejection on authentication attempt
- Test: revoked user cannot access any institution-authenticated routes
- Test: revocation is audited — audit record written
- Test: recovery is audited — audit record written
- Test: non-admin institution user cannot trigger recovery or revocation
- Test: step-up gate is enforced on recovery and revocation mutations

### Observability / telemetry
- Structured log for every recovery-trigger action (who initiated, for whom)
- Structured log for every revocation action (who initiated, for whom)
- These are security-sensitive actions — audit trail is mandatory

---

## Explicit in-scope deliverables

- Account recovery model: defined, implemented, and tested
- Hard revocation endpoint and backend model: implemented and enforced
- Revoked-user behavior: defined (cannot authenticate) and enforced in middleware
- Admin UX: recovery-trigger action visible and working
- Admin UX: hard revocation action with confirmation, visible and working
- Audit records for both recovery and revocation
- Named tests for all edge cases
- OpenAPI and contract docs updated for any new endpoints

---

## Explicit non-goals

- No sub-roles or scoped permission matrix (Phase H)
- No broad institution settings product expansion
- No redesign of the general authentication system (TOTP enrollment UX changes to enrolled state only)
- No hali-web work
- No citizen auth changes
- No recovery flows for citizen accounts

---

## Acceptance criteria

- [ ] Institution admin can trigger account recovery for a stranded institution user
- [ ] After recovery-trigger, the affected user must complete re-enrollment before gaining access
- [ ] Institution admin can hard-revoke a user's institution access (distinct from role demotion)
- [ ] Revoked user is rejected at the authentication boundary — receives a clear, non-misleading error
- [ ] Revoked user cannot reach any institution-authenticated route
- [ ] No privilege-escalation path is introduced (confirm: a revoked user cannot self-recover without admin action)
- [ ] Both recovery-trigger and revocation are auditable — audit records exist and are structured
- [ ] Step-up session gate is enforced on both actions
- [ ] Non-admin institution user cannot trigger recovery or revocation
- [ ] OpenAPI spec and contract docs are updated
- [ ] Named tests prove all behaviors listed above

---

## Dependencies

- Phase A: institution auth foundation (TOTP enrollment, session hydration) — **complete**
- Phase B: step-up session and `useStepUpRecovery` — **complete** (PR #335)
- Phase C: Members screen and admin write mutations — **complete** (PR #338)
- Phase E: operator UI surfaces — **should be complete or concurrent** (revocation UI builds on Members screen patterns)
- Audit logging infrastructure (PR #350) — **complete**

---

## Risks / pitfalls

- **Recovery model ambiguity**: the choice between TOTP-reset-by-admin vs backup codes has different UX and security trade-offs. Document the decision explicitly in the PR body before implementing.
- **Soft vs hard revocation confusion**: implementers may implement revocation as "role = null" or "role = viewer" instead of a distinct revoked state. This creates a silent gap — a viewer-role user still has access. Use a distinct, enforced revocation state.
- **Middleware bypass**: revoked-user rejection must happen at the auth middleware level, not just in the frontend. Test that revoked users cannot reach authenticated API routes.
- **Audit trail scope**: recovery and revocation are the highest-sensitivity admin actions on the platform. Audit trail is not optional.
- **Step-up bypass**: both actions must go through step-up gate. Verify `useStepUpRecovery` is wired, not assumed.
- **Schema migration risk**: if a new column or enum value is needed on `institution_memberships`, ensure the migration is safe and tested against real data shapes.

---

## Suggested implementation order

1. Define and document the recovery model (TOTP-reset-by-admin vs backup codes) — decide in PR body
2. Schema migration for revoked state on institution memberships (if needed)
3. Hard revocation backend endpoint and middleware enforcement
4. Recovery-trigger backend endpoint and TOTP reset logic
5. Audit records for both actions
6. Admin UI: revocation action + confirmation modal
7. Admin UI: recovery-trigger action
8. Step-up gate wiring for both
9. Error UX for revoked user
10. Named tests for all scenarios (revocation, recovery, non-admin block, revoked-user rejection)

---

## Notes for future execution prompt

- Specify: "decide and document recovery model in PR body before implementing"
- Specify: "hard revocation = distinct state, not role demotion — test revoked user cannot reach authenticated routes"
- Include: read `docs/arch/07_auth_implementation.md` before touching auth-related code
- Include: audit trail is mandatory for both actions — structured log + audit record
- Include: step-up gate required on both mutations
- Include: update OpenAPI spec and `AUTHORIZATION_COVERAGE_MATRIX.md` for any new protected routes
- Recommended sequence: schema → backend endpoints → middleware enforcement → admin UX → tests

### PR DESCRIPTION
**Pull request by irvinesunday**
Closes #355

## Summary

- **Backend**: New `DELETE /v1/institution-admin/users/{userId}` hard-revokes institution access (sets `IsBlocked`, terminates all active sessions). New `POST /v1/institution-admin/users/{userId}/reset-totp` clears the user's TOTP secret and recovery codes.
- **Login gate**: `VerifyMagicLink` now returns `auth.access_revoked` (401) for blocked accounts — blocked users cannot start a new session via magic link.
- **Frontend**: `RevokeAccessModal` with step-up recovery wired into `MembersScreen`. Blocked members show an "Access revoked" badge; action buttons are hidden for blocked users.
- **OpenAPI**: `DELETE /v1/institution-admin/users/{userId}` added as a second operation on the existing path. `POST /v1/institution-admin/users/{userId}/reset-totp` added. Four new error codes added to the enum.

## Test plan

- [x] `dotnet build` — 0 errors
- [x] `dotnet test tests/Hali.Tests.Unit` — 625/625 passing
- [x] `npx vitest run` — 252/252 passing
- [x] `npx tsc --noEmit` — clean
- [x] `ErrorCodeCatalogTests` passes (OpenAPI enum parity verified)
- [x] New frontend tests: `RevokeAccessModal.test.tsx` (5 cases) + `MembersScreen.test.tsx` extended (revoke/TOTP/blocked badge)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/Hali.Infrastructure/InstitutionAdmin/InstitutionAdminRepository.cs [hunk 1]
```diff
@@ -118,4 +118,51 @@ public async Task<IReadOnlyDictionary<Guid, string>> GetLocalityDisplayNamesAsyn
...
+    }
+
+    public async Task RevokeTotpAndClearRecoveryCodesAsync(
+        TotpSecret secret, Guid userId, DateTime revokedAt, CancellationToken ct)
+    {
+        secret.RevokedAt = revokedAt;
+        await _authDb.TotpRecoveryCodes
+            .Where(rc => rc.AccountId == userId && rc.UsedAt == null)
+            .ExecuteDeleteAsync(ct);
+        await _authDb.SaveChangesAsync(ct);
```
[the remark this answers is not in the document]
Comment (irvinesunday) [reply src/Hali.Infrastructure/InstitutionAdmin/InstitutionAdminRepository.cs]: Fixed in commit 63f4628. Wrapped both writes in `Database.CreateExecutionStrategy().ExecuteAsync` + an explicit `BeginTransactionAsync` / `CommitAsync`, exactly as suggested. The `ExecuteDeleteAsync` and `SaveChangesAsync` now share the same transaction, so a failure after the delete rolls back both.
